src/common/ad_tx.c

```diff
@@ -8,8 +8,193 @@
+	d_list_t	atc_tx_list;
+	d_list_t	atc_act_copy_list;
+	int		atc_act_nr;
+	int		atc_tx_nr;
+	int		atc_act_copy_nr;
+};
+
 static __thread struct ad_tx	*tls_tx;
 
+#define AD_TLS_CACHE_ENABLED	(1)
+#define TLS_ACT_NUM		(64)
+#define TLS_ACT_MAX		(512)
+#define TLS_TX_NUM		(16)
+#define TLS_ACT_COPY_NUM	(64)
+#define TLS_ACT_COPY_MAX	(256)
+/* payload size of cached UMEM_ACT_COPY act, if payload size exceed it then
+ * will directly allocate.
+ */
+#define TSL_ACT_COPY_SZ		(512)
+
+static __thread struct ad_tls_cache tls_cache;
+
+static struct ad_act *
+tls_act_get(int opc, size_t size)
+{
+	struct ad_act	*act;
+
+#if AD_TLS_CACHE_ENABLED
+	if (opc != UMEM_ACT_COPY && tls_cache.atc_act_nr > 0) {
+		act = d_list_pop_entry(&tls_cache.atc_act_list, struct ad_act, it_link);
+		act->it_act.ac_opc = opc;
+		tls_cache.atc_act_nr--;
+		return act;
+	}
+	if (opc == UMEM_ACT_COPY && size <= TSL_ACT_COPY_SZ && tls_cache.atc_act_copy_nr > 0) {
+		act = d_list_pop_entry(&tls_cache.atc_act_copy_list, struct ad_act, it_link);
+		act->it_act.ac_opc = opc;
+		tls_cache.atc_act_copy_nr--;
+		return act;
+	}
+#endif
+
+	if (opc == UMEM_ACT_COPY) {
+		size = max(size, TSL_ACT_COPY_SZ);
+		D_ALLOC_NZ(act, offsetof(struct ad_act, it_act.ac_copy.payload[size]));
+	} else {
+		D_ALLOC_PTR_NZ(act);
+	}
+	if (likely(act != NULL)) {
+		D_INIT_LIST_HEAD(&act->it_link);
+		act->it_act.ac_opc = opc;
+	}
+
+	return act;
+}
+
+static void
+tls_act_put(struct ad_act *act)
+{
+	d_list_del(&act->it_link);
+
+#if AD_TLS_CACHE_ENABLED
+	if (act->it_act.ac_opc == UMEM_ACT_COPY && act->it_act.ac_copy.size <= TSL_ACT_COPY_SZ &&
+	    tls_cache.atc_act_copy_nr < TLS_ACT_COPY_MAX) {
+		d_list_add(&act->it_link, &tls_cache.atc_act_copy_list);
+		tls_cache.atc_act_copy_nr++;
+		return;
+	}
+	if (act->it_act.ac_opc != UMEM_ACT_COPY && tls_cache.atc_act_nr < TLS_ACT_MAX) {
+		d_list_add(&act->it_link, &tls_cache.atc_act_list);
+		tls_cache.atc_act_nr++;
+		return;
+	}
+#endif
+
+	D_FREE(act);
+}
+
+static struct umem_wal_tx *
+tls_utx_get(void)
+{
+	struct umem_wal_tx_item	*item;
+
+#if AD_TLS_CACHE_ENABLED
+	if (tls_cache.atc_tx_nr > 0) {
+		item = d_list_pop_entry(&tls_cache.atc_tx_list, struct umem_wal_tx_item, ti_link);
+		item->ti_utx.utx_stage = 0;
+		tls_cache.atc_tx_nr--;
+		return &item->ti_utx;
+	}
+#endif
+	D_ALLOC_PTR(item);
+	if (item == NULL)
+		return NULL;
+	D_INIT_LIST_HEAD(&item->ti_link);
+	return &item->ti_utx;
+}
+
+static void
+tls_utx_put(struct umem_wal_tx *utx)
+{
+	struct umem_wal_tx_item	*item;
+
+	item = container_of(utx, struct umem_wal_tx_item, ti_utx);
+
+#if AD_TLS_CACHE_ENABLED
+	d_list_add(&item->ti_link, &tls_cache.atc_tx_list);
+	tls_cache.atc_tx_nr++;
+	return;
+#endif
+
+	D_FREE(item);
+}
+
+void
+ad_tls_cache_init(void)
+{
+	struct umem_wal_tx_item	*item;
+	struct ad_act		*act;
+	int			 i;
+
+	tls_cache.atc_act_nr = 0;
+	D_INIT_LIST_HEAD(&tls_cache.atc_act_list);
+	for (i = 0; i < TLS_ACT_NUM; i++) {
+		D_ALLOC_PTR(act);
+		if (act == NULL)
+			return;
```

Comment:
error code?

Reply:
intended to not return err code to simplify err handing. as the act pre-post is not must needed.